src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc

```diff
@@ -0,0 +1,190 @@
+    PrintIndent();
+    ss << "#ifdef __cplusplus\n";
+    ss << "extern \"C\" {\n";
+    ss << "#endif\n";
+  }
+
+  /*!  * \brief Creates a cplusplus guard postfix for extern "C" printing */
+  void PrintExternCPostfix(std::ostringstream& ss) {
+    PrintIndent();
+    ss << "#ifdef __cplusplus\n";
+    ss << "}\n";
+    ss << "#endif\n";
+  }
+
+  /*!  * \brief Emits CMSIS-NN code block for softmax */
+  void EmitSoftmax(const CallNode* op) {
+    // @tir.call_extern("arm_softmax_s8", buffer_0, num_rows, row_size,
+    //                  mult, shift, diff_min, buffer_1, dtype=int8)
+    std::string cmsis_func_name = op->args[0].as<StringImmNode>()->value;
+    int32_t num_rows = op->args[2].as<IntImmNode>()->value;
+    int32_t row_size = op->args[3].as<IntImmNode>()->value;
+    int32_t mult = op->args[4].as<IntImmNode>()->value;
+    int32_t shift = op->args[5].as<IntImmNode>()->value;
+    int32_t diff_min = op->args[6].as<IntImmNode>()->value;
+
+    PrintIndent();
+    stream << "int32_t num_rows = " << num_rows << ";\n";
+    PrintIndent();
+    stream << "int32_t row_size = " << row_size << ";\n";
+    PrintIndent();
+    stream << "int32_t mult = " << mult << ";\n";
+    PrintIndent();
+    stream << "int32_t shift = " << shift << ";\n";
+    PrintIndent();
+    stream << "int32_t diff_min = " << diff_min << ";\n";
+    PrintIndent();
+    stream << cmsis_func_name << "(input,";
+    PrintIndent();
+    stream << " num_rows, row_size, mult, shift, diff_min, output);\n";
+    PrintIndent();
+    stream << "return;\n";
+  }
+};
+
+class CMSISNNModuleNode : public runtime::ModuleNode {
+ public:
+  CMSISNNModuleNode(const std::string& code, const std::string& fmt,
+                    const Array<String>& func_names)
+      : code_(code), fmt_(fmt), func_names_(func_names) {}
+
+  std::string GetSource(const std::string& format) final { return code_; }
+
+  const char* type_key() const { return "c"; }
+
+  PackedFunc GetFunction(const std::string& name, const ObjectPtr<Object>& sptr_to_self) final {
+    if (name == "get_symbol") {
+      return PackedFunc(
+          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->func_names_[0]; });
+    } else if (name == "get_func_names") {
+      return PackedFunc(
+          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->func_names_; });
+    } else {
+      return PackedFunc(nullptr);
+    }
+  }
+
+  void SaveToFile(const std::string& file_name, const std::string& format) final {
+    std::string fmt = GetFileFormat(file_name, format);
+    std::string meta_file = GetMetaFilePath(file_name);
+    if (fmt == "c" || fmt == "cu") {
+      ICHECK_NE(code_.length(), 0);
+      SaveBinaryToFile(file_name, code_);
+    } else {
+      ICHECK_EQ(fmt, fmt_) << "Can only save to format=" << fmt_;
+    }
+  }
+
+ protected:
+  std::string code_;
+  std::string fmt_;
+  Array<String> func_names_;
+};
+
+class CMSISNNModule : public Module {
```

Comment:
Where is this used ?

Reply:
I dont think this is exposed to python, because CMSISNNModuleNodeCreate uses the base runtime.Module class that wraps CMSISNNModuleNode. Are you able to access the ObjectRef that is CMSISNNModule in python ?

Reply:
You are right. CMSISNNModule got left out because later I followed in the suite of codegen_c_host.cc to create runtime.Module directly. Before that, when CMSISNNModule was getting returned, iirc, code was faulting somewhere in compile_engine.h. Since it is of no use now, we can remove it.